### DISCUSSION TITLE
Idiomatic way for manipulating a &ByteStream

### DISCUSSION THREAD

**Discussion by vhsconnect** [awslabs/aws-sdk-rust#1044] [Q&A] [answered]
Hello,

I am trying to write a little wrapper library around the `aws_sdk_glacier` module. I am having troubles manipulating the reference to ByteStream from calling `get_job_output` - as it a shared reference and I am unable to move it. Can't seem to call `collect` or `into_inner` and write those bytes to a file.

Below is a small example

```rust
async fn write_job_output(client: &Client) -> Result<(), Box<dyn std::error::Error>> {
    let request = client
        .get_job_output()
        .account_id("...")
        .vault_name("...")
        .job_id("...")
        .send();
    let request = request.await?;

    let mut file = File::create("output.json").await?;
    file.write_all(request.body().into_inner().bytes().unwrap()).await?; // error here: cannot move out of a shared reference

    Ok(())
}
```

**Comment by rcoh** [comment, marked as answer]
Use `.body` (without parens) to take ownership. I think you probably want `.collect`, or to iterate over the chunks. See example 2 here: https://docs.rs/aws-sdk-glacier/latest/aws_sdk_glacier/primitives/struct.ByteStream.html#getting-data-out-of-a-bytestream

**Reply by vhsconnect** [reply]
Thanks @rcoh, yes that works. 
Now it seems obvious looking at my LSP and the docs. I didn't know a Struct can have a field and a method of the same name - so I assumed I was looking at one of the same the whole time.

Out of curiosity what would be the use case of using the method variant of `body()` or `checksum()`

**Reply by rcoh** [reply]
you could introspect things like content length. They aren't super useful. It would be nice to have docs on the field pointing you to the named field, honestly.
